include/envoy/stream_info/filter_state.h

```diff
@@ -55,6 +55,69 @@ class FilterState {
             (dynamic_cast<const T*>(getDataGeneric(data_name)) != nullptr));
   }
 
+  /**
+   * The addToList, hasListWithName, forEachListItem are variants of the above
+   * functions, that operate on list data. Multiple elements could be added
+   * the list under the same data_name.
+   * @param data_name the name of the data being added to the end of a list.
+   * @param data an owning pointer to the data to be stored.
+   * Note that data_names for list elements have no relation to the data_names
+   * for singleton data objects added through setData. All items added to the list
+   * must be of the same type.
+   */
+  template <typename T>
+  void addToList(absl::string_view data_name, std::unique_ptr<Object>&& data) {
+    const auto list = getList(data_name);
+    if (list != nullptr) {
+      // Check type of first element in the list
+      const T* cast = dynamic_cast<const T*>(list->at(0).get());
+      if (!cast) {
+        throw EnvoyException(
+            fmt::format("List {} does not conform to the specified type", data_name));
+      }
+    }
+
+    addToListGeneric(data_name, std::move(data));
+  }
+
+  /**
+   * @param data_name the name of the list being probed.
+   * @return Whether a list of the type and name specified exists in the
+   * data store.
+   */
+  template <typename T> bool hasList(absl::string_view data_name) const {
+    const auto list = getList(data_name);
+    return ((list != nullptr) && (dynamic_cast<const T*>(list->at(0).get()) != nullptr));
```

Comment:
HasList is supposed to be non-punishing, in that its the only way to check if something has an object with that type. The exception throwing bits are in getData, setData, forEachList, etc.. (i.e. without something to allow you to check easily, it gets very clunky to even to presence checks).

Reply:
I guess the way I look at it, it is always a compile-time error to look for a given key with the wrong type. In other words, every reference to a given key should use the exact same type. So to aid in testing, making the code-writer aware asap is useful.

Reply:
hmm.. But doesn't this create an inconsistent interface with hasData<> that has a similar behavior in that it doesn't throw an error if the item can't be type casted? I can change hasData as well in that case, though I don't know if Google folks will have issues with breaking this interface.

Reply:
Also, I could always do a try catch in my code, even if we throw error from hasList.. At that time, we are essentially pushing boilerplate stuff to filter authors that want to check if someone already created a list of a specific type, if so read it, append "if business logic says so", and move on..

The error throwing part will now force people to either do try catch stuff or blindly append to the list [which I sometimes may not want to, depending on the protocol codecs in the filter chain. E.g., an oracle auth protocol decoder filter reads something and sets attributes under <OracleSourceIdentity> type and key srcidentity. then the oracle data stream protocol decoder reads the protocol payload and finds sourceIdentity. It would do hasList<OracleAttributes>(srcidentity). if it exists, leave it as is (i.e. priority given to the other codec). If not, add these attributes..

Reply:
@curiouserrandy do you have an opinion on this?

Reply:
So my original design was "both/and" in that the user could use hasDataWithName() to determine if something existed with any type, and hasData<> to determine if it existed with a particular type.  I'm not wedded to that design, but it would be nice to be parallel between the two mechanisms.   (Though this depends on whether we unify the namespaces as I suggest elsewhere--if we don't, we need a non-throwing interface to probe each store.)

Having said that, I do agree that if a name is used in two places in the code and different types are used, that's a mistake, and making the author aware of it ASAP is a win.  So I wouldn't object to making the two mechanisms equivalent by yanking hasDataWithName() from the original interface.  If we go based on the rule of thumb to not add interfaces unless we have use cases for it, that would probably be the right way to go.

@rshriram : On that line, I'm confused about the use case you describe.  In that use case, aren't all code locations aware of the type involved and hence could use hasList<> without having catches?